working/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,470 @@
+    library with a new class that calls that private generative constructor from
+    its own public one. That subclass is now externally extensible and the
+    language quietly lets you do that.
+
+*   **Trust but verify.** In the above example, it's not clear what the author
+    *intends*. Maybe they deliberately didn't put any modifiers on `MySubclass`
+    because they *want* to re-add the capability that its superclass removed.
+    But maybe they just didn't notice that `NoExtend` removed them, or they
+    forgot to put `interface` on `MySubclass`.
+
+    Since it's not clear what they meant, the language could require them to
+    clarify. If you define a subtype of a type that has removed a capability, we
+    could require you to annotate specifically when you re-add that capability.
+    If you don't intend to re-add a capability, you restate the restriction:
+
+    ```dart
+    interface class NoExtend {}
+    interface class MySubclass extends NoExtend {}
+    ```
+
+    And if you *do* intend to loosen it, you make that explicit by some marker
+    like:
+
+    ```dart
+    interface class NoExtend {}
+    reopen class MySubclass extends NoExtend {}
+    ```
+
+    Here "reopen" means, "I know I didn't put any other modifier here and that
+    means this class has more capabilities than my parent."
+
+    Personally, I think this is probably more modifiers than we want and is
+    more trouble than it's worth. I worry about having to explain to users that
+    a class marked `reopen` means the same thing as a class not marked with it.
+    But I do think it could be useful to offer this as a lint with a metadata
+    annotation for users that are more cautious, like:
+
+    ```dart
+    interface class NoExtend {}
+    @reopen class MySubclass extends NoExtend {}
+    ```
+
+    This proposal takes that option.
+
+Finally, to the actual proposal...
+
+## Syntax
+
+This proposal builds on the existing sealed types proposal so the grammar
+includes those changes. The full set of modifiers that can appear before a class
+or mixin are `abstract`, `sealed`, `base`, `interface`, and `final`. Some
+combinations don't make sense:
+
+*   `sealed` implies `abstract`, so they can't be combined.
+*   `sealed` implies non-extensibility, so can't be combined with `interface`
+    or `final`.
+*   `sealed` implies non-implementability, so can't be combined with `base` or
+    `final`.
+*   `base`, `interface`, and `final` all control the same two capabilities so
+    are mutually exclusive.
+*   `final` and `interface` imply that the type can't be used as a superclass,
+    but the reason for defining a mixin is so that it can be mixed in and become
+    a superclass, so we don't allow those on mixins.
+
+The remaining valid combinations are:
+
+```
+class
+sealed class
+base class
+interface class
+final class
+abstract class
+abstract base class
+abstract interface class
+abstract final class
```

Comment:
So I'd add:

* `mixin class`
* `sealed mixin class`
* `base mixin class`
* `abstract mixin class`
* `abstract base mixin class`

as well. Those are the only classes which can be mixed in. (Possibly only outside of the library.)